### PR TITLE
[Ramda] fix and improve propOr

### PR DESCRIPTION
**Pull request by valerii15298**
Currently, typings for `propOr` are wrong and do not support curing.
I tried to add most broad definition for all use cases.

In `PO` I used `NonNullable` because ramda will use defaultValue if value by key is undefined or null.
Also use name `PO` because it is used so often in definitions that code becames bloated. So I choose small name for it.
Happy to hear suggestions.

### CODE REVIEW

File: types/ramda/index.d.ts [hunk 1]
```diff
@@ -2331,11 +2333,58 @@ export function propIs<C extends new (...args: any[]) => any>(
...
+    defaultValue: Placeholder,
+    propName: P,
+    obj: O,
+): <D>(defaultValue: D) => D | PO<P, O>;
+export function propOr<D, O extends object>(
+    defaultValue: D,
+    propName: Placeholder,
+    obj: O,
+): <P extends ObjectKeyType>(propName: P) => D | PO<P, O>;
+export function propOr<D, P extends ObjectKeyType, O extends object>(
```
Comment (valerii15298) [inline types/ramda/index.d.ts:2358]: Here is the main definition and logic. All others are just for curring.

File: types/ramda/tools.d.ts [hunk 2]
```diff
@@ -408,6 +408,24 @@ export type ObjPred<T = unknown> = (value: any, key: unknown extends T ? string
  */
 export type Ord = number | string | boolean | Date;
 
+/*
```
Comment (adispring) [inline types/ramda/tools.d.ts:411]: Please use [jsdoc style comment](https://google.github.io/styleguide/tsguide.html#jsdoc-vs-comments).

Can you modify other comments introduced before to jsdoc style comment?

**Review by adispring — APPROVED** [review]

File: types/ramda/index.d.ts [hunk 3]
```diff
@@ -2331,11 +2331,58 @@ export function propIs<C extends new (...args: any[]) => any>(
  * If the given, non-null object has an own property with the specified name, returns the value of that property.
  * Otherwise returns the provided default value.
  */
-export function propOr<T, U>(val: T, __: Placeholder, obj: U): <V>(p: string) => V;
-export function propOr<U>(__: Placeholder, p: string, obj: U): <T, V>(val: T) => V;
-export function propOr<T, U, V>(val: T, p: string, obj: U): V;
-export function propOr<T>(val: T, p: string): <U, V>(obj: U) => V;
-export function propOr<T>(val: T): <U, V>(p: string, obj: U) => V;
+export function propOr<O extends object>(
```
Comment (adispring) [inline types/ramda/index.d.ts:2334]: Placeholders introduce too much complexity with little gains, so I don't think it's a good idea to continue compatibility with placeholders.

**Review by adispring — CHANGES_REQUESTED** [review]

File: types/ramda/index.d.ts (hunk 3 above)
Comment (essenmitsosse) [reply types/ramda/index.d.ts:2334]: I also wouldn't add those typings until there is a decision wether or not we want to continue support

**Comment by valerii15298** [comment]
@jakebailey can you give advice please? CI uses old version of typescript 4.7 while in this repo typescript version is set to `next` and with 4.8 all tests are good when I run them locally.

**Comment by jakebailey** [comment]
If you're talking about the union ordering issues, you can write both versions for each and separate them with `||`.

IIRC we're working on making sort ordering more consistent in future versions of TS to alleviate this.

**Pull request closed without merging**